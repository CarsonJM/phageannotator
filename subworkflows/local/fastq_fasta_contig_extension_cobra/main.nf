//
// Extend contig ends using COBRA
//

include { BOWTIE2_BUILD                 } from '../../../modules/nf-core/bowtie2/build/main'
include { FASTQ_ALIGN_BOWTIE2           } from '../../nf-core/fastq_align_bowtie2/main'
include { GUNZIP                        } from '../../../modules/nf-core/gunzip/main'
include { COVERM_CONTIG as COVERM_COBRA } from '../../../modules/local/coverm/contig/main'
include { COBRA                         } from '../../../modules/local/cobra/main'

workflow FASTQ_FASTA_CONTIG_EXTENSION_COBRA {
    take:
    fastq_gz            // [ [ meta ], [ reads_1.fastq.gz, reads_2.fastq.gz ] ] , read files (mandatory)
    fasta_gz            // [ [ meta ], fasta.gz ]                               , assemblies/scaffolds (mandatory)
    viral_contigs_tsv   // [ [ meta ], virus_contigs.tsv ]                      , TSV file containing viral contig names

    main:
    ch_versions = Channel.empty()

    //
    // MODULE: Make bowtie2 index
    //
    ch_fasta_bt2 = BOWTIE2_BUILD ( fasta_gz ).index.first()
    ch_versions = ch_versions.mix( BOWTIE2_BUILD.out.versions )

    //
    // SUBWORKFLOW: Align reads to bowtie2 index
    //
    ch_fasta_alignment_bam = FASTQ_ALIGN_BOWTIE2 ( fastq_gz, ch_fasta_bt2, false, false, fasta_gz ).bam
    ch_versions = ch_versions.mix( FASTQ_ALIGN_BOWTIE2.out.versions )

    //
    // MODULE: Calculate abundance metrics from BAM file
    //
    ch_coverage_tsv = COVERM_COBRA ( ch_fasta_alignment_bam ).alignment_results
    ch_versions = ch_versions.mix( COVERM_COBRA.out.versions )

    // Remove header from
    ch_coverage_mod = ch_coverage_tsv
    .map { meta, tsv ->
        def tsv_mod = file("${workDir}/${meta.id}_coverage_mod.tsv")

        tsv.withReader { source ->
            tsv_mod.withWriter { target ->
                String line
                while( line=source.readLine() ) {
                    if ( !line.startsWith('Contig' )) {
                        target << line << '\n'
                    }
                }
            }
        }

        return [ meta, tsv_mod]
    }

    //
    // MODULE: Gunzip FASTA file
    //
    ch_fasta = GUNZIP ( fasta_gz ).gunzip
    ch_versions = ch_versions.mix( GUNZIP.out.versions )

    // prepare input for cobra
    ch_cobra_input = ch_fasta
        .join(ch_coverage_mod)
        .join(viral_contigs_tsv)
        .join(ch_fasta_alignment_bam)


    //
    // MODULE: Extend contigs using COBRA
    //
    ch_extended_fasta_gz = COBRA ( ch_cobra_input, params.cobra_assembler, params.cobra_mink, params.cobra_maxk ).extended_assemblies
    ch_versions = ch_versions.mix( COBRA.out.versions )

    emit:
    extended_fasta_gz   = ch_extended_fasta_gz  // [ [ meta ], extended_contigs.fna.gz ]    , FASTA file containing extended contigs
    versions            = ch_versions.unique()  // [ versions.yml ]
}
