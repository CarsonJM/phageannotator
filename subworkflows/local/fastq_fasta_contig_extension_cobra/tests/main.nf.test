nextflow_workflow {

    name "Test Subworkflow: FASTQ_FASTA_CONTIG_EXTENSION_COBRA"
    script "subworkflows/local/fastq_fasta_contig_extension_cobra/main.nf"
    workflow "FASTQ_FASTA_CONTIG_EXTENSION_COBRA"
    tag "subworkflows"
    tag "subworkflows_local"
    tag "fastq_fasta_contig_extension_cobra"
    tag "fastq_fasta_contig_extension_cobra_default"


    test("Parameters: default") {

        when {
            params {
                outdir              = "$outputDir"
                cobra_assembler     = 'metaspades'
                cobra_mink          = 21
                cobra_maxk          = 127
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'test' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkifExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkifExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id: 'test' ],
                        file("$projectDir/assets/test_data/subworkflows/fastq_fasta_contig_extension_cobra/sars_cov2_contigs.fasta.gz", checkifExists: true)
                    ]
                )
                input[2] = Channel.of(
                    [
                        [ id: 'test' ],
                        file("$projectDir/assets/test_data/subworkflows/fastq_fasta_contig_extension_cobra/viral_contigs.tsv", checkifExists: true)
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
