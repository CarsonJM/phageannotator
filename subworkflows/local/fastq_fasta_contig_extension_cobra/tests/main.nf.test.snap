{
    "Parameters: default": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_extended.fasta.gz:md5,0b52beff348887e1552c9c956ba53020"
                    ]
                ],
                "1": [
                    "versions.yml:md5,2bcdeb17ef38b559ef7ab4d3284cec38",
                    "versions.yml:md5,3d80b62c3584609643c289596649c76d",
                    "versions.yml:md5,45ab424acfec849d28d8f8ab4a3af8f1",
                    "versions.yml:md5,466c764b0ff156331507ba238e111bf2",
                    "versions.yml:md5,4ee89a120decfaf19bf7a9b274485ba4",
                    "versions.yml:md5,64da2eecb8de1b06695a1d90486247b0",
                    "versions.yml:md5,78a67cf0bf546994c775283ac774760a",
                    "versions.yml:md5,7f61c2e84c088e886fef2f98710466b1",
                    "versions.yml:md5,d464199cdf99a00939311c178e24c614",
                    "versions.yml:md5,fc102771b3196a7b96fec4a6f1281a37"
                ],
                "extended_fasta_gz": [
                    [
                        {
                            "id": "test"
                        },
                        "COBRA_extended.fasta.gz:md5,0b52beff348887e1552c9c956ba53020"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2bcdeb17ef38b559ef7ab4d3284cec38",
                    "versions.yml:md5,3d80b62c3584609643c289596649c76d",
                    "versions.yml:md5,45ab424acfec849d28d8f8ab4a3af8f1",
                    "versions.yml:md5,466c764b0ff156331507ba238e111bf2",
                    "versions.yml:md5,4ee89a120decfaf19bf7a9b274485ba4",
                    "versions.yml:md5,64da2eecb8de1b06695a1d90486247b0",
                    "versions.yml:md5,78a67cf0bf546994c775283ac774760a",
                    "versions.yml:md5,7f61c2e84c088e886fef2f98710466b1",
                    "versions.yml:md5,d464199cdf99a00939311c178e24c614",
                    "versions.yml:md5,fc102771b3196a7b96fec4a6f1281a37"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-02-28T11:22:56.392813058"
    }
}